File: assets/experiment_score/calculate_fnv1a.py
#!/usr/bin/env python3

import sys
import json
from pathlib import Path
from PIL import Image


def calculate_pixel_fnv1a(file_path):
    image = Image.open(file_path).convert("RGBA")
    
    pixel_bytes = []
    pixels = image.load()
    
    for y in range(20):
        for x in range(50):
            pixel = pixels[x, y]
            r, g, b, a = pixel[0], pixel[1], pixel[2], pixel[3]
            
            if a == 255:
                pixel_bytes.extend([r, g, b])
    
    hash_value = 0x811C9DC5
    for byte in pixel_bytes:
        hash_value ^= byte
        hash_value = (hash_value * 0x01000193) & 0xFFFFFFFF
    
    return hash_value


directory_path = sys.argv[1]
output_path = sys.argv[2] if len(sys.argv) > 2 else "score_hashes.json"

directory = Path(directory_path)
file_hashes = {}

for file_path in directory.rglob('*'):
    if file_path.suffix.lower() in {'.png', '.jpg', '.jpeg', '.bmp', '.gif'}:
        relative_path = file_path.relative_to(directory)
        file_key = str(relative_path.with_suffix(''))
        
        print(f"Processing: {relative_path}")
        hash_value = calculate_pixel_fnv1a(str(file_path))
        file_hashes[file_key] = hash_value

with open(output_path, 'w', encoding='utf-8') as f:
    json.dump(file_hashes, f, ensure_ascii=False, separators=(',', ':'))

print(f"\nProcessed {len(file_hashes)} files")
print(f"Saved to: {Path(output_path).absolute()}")

File: assets/experiment_score/calculate_fnv1a.dart
// ignore_for_file: avoid_print

import 'dart:convert';
import 'dart:io';
import 'package:image/image.dart' as img;
import 'package:path/path.dart' as path;

void main(List<String> arguments) async {
  // 检查是否提供了文件夹路径参数
  if (arguments.isEmpty) {
    print('请提供文件夹路径作为参数');
    print('用法: dart md5_calculator.dart <文件夹路径> [输出JSON路径]');
    return;
  }

  // 获取输入文件夹路径
  final directoryPath = arguments[0];
  final directory = Directory(directoryPath);

  // 检查文件夹是否存在
  if (!await directory.exists()) {
    print('错误: 文件夹 "$directoryPath" 不存在');
    return;
  }

  // 确定输出文件路径（默认在当前目录生成md5_hashes.json）
  final outputPath = arguments.length > 1 ? arguments[1] : 'score_hashes.json';

  try {
    // 计算所有文件的哈希值并收集结果
    final Map<String, int> fileHashes = {};
    await for (final entity in directory.list(recursive: true)) {
      if (entity is File) {
        // 获取相对路径作为文件名标识
        final relativePath = path.relative(entity.path, from: directoryPath);
        print('正在计算: $relativePath');

        // 计算哈希值
        final hash = await _calculatePixelFNV1A(entity);
        fileHashes[relativePath.substring(0, relativePath.length - 4)] = hash;
      }
    }

    // 保存结果到JSON文件
    final outputFile = File(outputPath);
    await outputFile.writeAsString(
      json.encode(fileHashes, toEncodable: (obj) => obj.toString()),
      mode: FileMode.write,
    );

    print('计算完成，共处理 ${fileHashes.length} 个文件');
    print('结果已保存到: ${outputFile.absolute.path}');
  } catch (e) {
    print('处理过程中发生错误: $e');
  }
}

/// 计算单个文件的FNV哈希值
Future<int> _calculatePixelFNV1A(File file) async {
  final bytes = file.readAsBytesSync();

  // 解码图片为 image 库的 Image 对象
  final image = img.decodeImage(bytes);
  if (image == null) {
    throw Exception("img.decodeImage(bytes) return null. Unsupported format.");
  }

  // 遍历像素，提取RGB通道
  final width = 50;
  final height = 20;
  final pixelBytes = <int>[];

  for (int y = 0; y < height; y++) {
    for (int x = 0; x < width; x++) {
      final pixel = image.getPixel(x, y);

      final a = pixel.a.toInt();
      if (a == 255) {
        // ignore the transparent pixel to reduce the calculation
        pixelBytes.addAll([pixel.r.toInt(), pixel.g.toInt(), pixel.b.toInt()]);
      }
    }
  }

  var hash = 0x811C9DC5;
  for (var p in pixelBytes) {
    hash ^= p;
    hash = (hash * 0x01000193) & 0xFFFFFFFF;
  }

  return hash;
}
